### PR TITLE
KTOR-1094 Merge abstract base with interfaces in IO

### PR DESCRIPTION
**Pull request by e5l**

### CODE REVIEW

File: ktor-io/jvm/src/io/ktor/utils/io/charsets/CharsetJVM.kt [hunk 1]
```diff
@@ -63,7 +63,7 @@ public actual fun CharsetEncoder.encodeUTF8(input: ByteReadPacket, dst: Output)
 
     try {
         tmp.writeDirect(0) { tmpBb ->
-            val cb = tmpBb.asCharBuffer()!!
+            val cb = tmpBb.asCharBuffer()
```
Comment (cy6erGn0m) [inline ktor-io/jvm/src/io/ktor/utils/io/charsets/CharsetJVM.kt:66]: This actually makes difference as removing it causes compiled to add assert to every every variable access. This is why I do like this to make this check only once.

File: ktor-io/jvm/src/io/ktor/utils/io/charsets/CharsetJVM.kt [hunk 2]
```diff
@@ -147,7 +147,7 @@ internal actual fun CharsetDecoder.decodeBuffer(
     var charactersCopied = 0
     input.readDirect { bb ->
         val tmpBuffer = ChunkBuffer.Pool.borrow()
-        val cb = tmpBuffer.memory.buffer.asCharBuffer()!!
+        val cb = tmpBuffer.memory.buffer.asCharBuffer()
```
Comment (cy6erGn0m) [inline ktor-io/jvm/src/io/ktor/utils/io/charsets/CharsetJVM.kt:150]: same here

File: ktor-io/jvm/src/io/ktor/utils/io/charsets/CharsetJVM.kt [hunk 3]
```diff
@@ -339,5 +339,5 @@ actual constructor(message: String) : java.nio.charset.MalformedInputException(0
         get() = _message
 }
 
-private val EmptyCharBuffer = CharBuffer.allocate(0)!!
+private val EmptyCharBuffer = CharBuffer.allocate(0)
```
Comment (cy6erGn0m) [inline ktor-io/jvm/src/io/ktor/utils/io/charsets/CharsetJVM.kt:342]: same

**Review by cy6erGn0m — APPROVED** [review]
please check `!!` impact: most likely we get slowdown because of this

File: ktor-io/jvm/src/io/ktor/utils/io/charsets/CharsetJVM.kt (hunk 1 above)
Comment (e5l) [reply ktor-io/jvm/src/io/ktor/utils/io/charsets/CharsetJVM.kt:66]: It marked with `@NotNull`, there is no platform types any more

**Pull request merged**